File: ports/ruby/lib/unicode_ruby/security/covert/bidi_control_balance.rb
# frozen_string_literal: true

module UnicodeRuby
  module Security
    module Covert
      # Detection of Trojan-Source-class bidi-control balance hazards
      # (CVE-2021-42574 / CVE-2021-42694).  Walks the input with per-type
      # stacks and produces four independent sub-threats.
      module BidiControlBalance
        Verdict = Struct.new(
          :kind, :sub, :bidi_positions,
          :emb_open_count, :emb_pop_count, :iso_open_count, :iso_pop_count,
          :max_depth
        )

        # UAX #9 §3.3.2 cap on the stack-of-stacks depth.
        UAX_DEPTH_LIMIT = 125

        module_function

        # LRE (202A), RLE (202B), LRO (202D), RLO (202E).
        def opens_embedding?(cp)
          cp == 0x202A || cp == 0x202B || cp == 0x202D || cp == 0x202E
        end

        def pdf?(cp)
          cp == 0x202C
        end

        # LRI (2066), RLI (2067), FSI (2068).
        def opens_isolate?(cp)
          cp == 0x2066 || cp == 0x2067 || cp == 0x2068
        end

        def pdi?(cp)
          cp == 0x2069
        end

        def bidi_format_control?(cp)
          opens_embedding?(cp) || pdf?(cp) || opens_isolate?(cp) || pdi?(cp)
        end

        def detect(input)
          v = Verdict.new(Calculus::ClassificationKind::CLEAR, nil, [], 0, 0, 0, 0, 0)
          emb_stack = 0
          iso_stack = 0
          orphans = []

          input.each_with_index do |cp, i|
            next unless bidi_format_control?(cp)

            v.bidi_positions << i
            if opens_embedding?(cp)
              emb_stack += 1
              v.emb_open_count += 1
              depth = emb_stack + iso_stack
              v.max_depth = depth if depth > v.max_depth
            elsif pdf?(cp)
              v.emb_pop_count += 1
              if emb_stack > 0
                emb_stack -= 1
              else
                orphans << i
              end
            elsif opens_isolate?(cp)
              iso_stack += 1
              v.iso_open_count += 1
              depth = emb_stack + iso_stack
              v.max_depth = depth if depth > v.max_depth
            elsif pdi?(cp)
              v.iso_pop_count += 1
              if iso_stack > 0
                iso_stack -= 1
              else
                orphans << i
              end
            end
          end

          return v if v.bidi_positions.empty?

          if v.max_depth > UAX_DEPTH_LIMIT
            v.kind = Calculus::ClassificationKind::HAZARD
            v.sub = "DepthExceeded"
            return v
          end
          unless orphans.empty?
            v.kind = Calculus::ClassificationKind::HAZARD
            v.sub = "OrphanPop"
            return v
          end
          if emb_stack > 0
            v.kind = Calculus::ClassificationKind::HAZARD
            v.sub = "UnbalancedEmbedding"
            return v
          end
          if iso_stack > 0
            v.kind = Calculus::ClassificationKind::HAZARD
            v.sub = "UnbalancedIsolate"
            return v
          end
          v
        end
      end
    end
  end
end

File: ports/elixir/lib/unicode_security/covert/bidi_control_balance.ex
defmodule UnicodeSecurity.Covert.BidiControlBalance do
  defstruct kind: :clear,
            sub: nil,
            bidi_positions: [],
            emb_open_count: 0,
            emb_pop_count: 0,
            iso_open_count: 0,
            iso_pop_count: 0,
            max_depth: 0

  @uax_depth_limit 125

  def opens_embedding?(cp), do: cp in [0x202A, 0x202B, 0x202D, 0x202E]
  def pdf?(cp), do: cp == 0x202C
  def opens_isolate?(cp), do: cp in [0x2066, 0x2067, 0x2068]
  def pdi?(cp), do: cp == 0x2069

  def bidi_format_control?(cp),
    do: opens_embedding?(cp) or pdf?(cp) or opens_isolate?(cp) or pdi?(cp)

  def detect(input) do
    init = {%__MODULE__{}, 0, 0, []}

    {v, emb_stack, iso_stack, orphans} =
      input
      |> Enum.with_index()
      |> Enum.reduce(init, fn {cp, i}, {v, emb, iso, orphans} ->
        if bidi_format_control?(cp) do
          v = %{v | bidi_positions: v.bidi_positions ++ [i]}

          cond do
            opens_embedding?(cp) ->
              emb = emb + 1

              {%{
                 v
                 | emb_open_count: v.emb_open_count + 1,
                   max_depth: max(v.max_depth, emb + iso)
               }, emb, iso, orphans}

            pdf?(cp) ->
              if emb > 0 do
                {%{v | emb_pop_count: v.emb_pop_count + 1}, emb - 1, iso, orphans}
              else
                {%{v | emb_pop_count: v.emb_pop_count + 1}, emb, iso, orphans ++ [i]}
              end

            opens_isolate?(cp) ->
              iso = iso + 1

              {%{
                 v
                 | iso_open_count: v.iso_open_count + 1,
                   max_depth: max(v.max_depth, emb + iso)
               }, emb, iso, orphans}

            pdi?(cp) ->
              if iso > 0 do
                {%{v | iso_pop_count: v.iso_pop_count + 1}, emb, iso - 1, orphans}
              else
                {%{v | iso_pop_count: v.iso_pop_count + 1}, emb, iso, orphans ++ [i]}
              end
          end
        else
          {v, emb, iso, orphans}
        end
      end)

    cond do
      v.bidi_positions == [] ->
        v

      v.max_depth > @uax_depth_limit ->
        %{v | kind: :hazard, sub: %{tag: "DepthExceeded", max_depth: v.max_depth}}

      orphans != [] ->
        %{v | kind: :hazard, sub: %{tag: "OrphanPop", positions: orphans}}

      emb_stack > 0 ->
        %{
          v
          | kind: :hazard,
            sub: %{
              tag: "UnbalancedEmbedding",
              open_count: v.emb_open_count,
              pop_count: v.emb_pop_count
            }
        }

      iso_stack > 0 ->
        %{
          v
          | kind: :hazard,
            sub: %{
              tag: "UnbalancedIsolate",
              open_count: v.iso_open_count,
              pop_count: v.iso_pop_count
            }
        }

      true ->
        v
    end
  end
end
